code/game/machinery/shieldgen.dm

```diff
@@ -553,3 +553,668 @@
+	icon_state = "charger_off_closed"
+
+	circuit = /obj/item/circuitboard/machine/modular_shield_charger
+
+	///Amount of regeneration this machine grants the connected generator
+	var/charge_boost = 0
+
+/obj/machinery/modular_shield/module/charger/update_icon_state()
+	. = ..()
+	if(isnull(shield_generator) || (machine_stat & NOPOWER))
+		icon_state = "charger_off_[panel_open ? "open" : "closed"]"
+		return
+	icon_state = "charger_on_[panel_open ? "open" : "closed"]"
+
+/obj/machinery/modular_shield/module/charger/RefreshParts()
+	. = ..()
+	charge_boost = initial(charge_boost)
+	for(var/datum/stock_part/servo/new_servo in component_parts)
+		charge_boost += new_servo.tier
+
+	if(shield_generator)
+		shield_generator.calculate_boost()
+
+/obj/machinery/modular_shield/module/relay
+
+	name = "Modular Shield Relay"
+	desc = "It helps the shield generator project farther out."
+	icon = 'icons/obj/machines/modular_shield_generator.dmi'
+	icon_state = "relay_off_closed"
+
+	circuit = /obj/item/circuitboard/machine/modular_shield_relay
+
+	///Amount of max range this machine grants the connected generator
+	var/range_boost = 0
+
+/obj/machinery/modular_shield/module/relay/update_icon_state()
+	. = ..()
+	if(isnull(shield_generator) || (machine_stat & NOPOWER))
+		icon_state = "relay_off_[panel_open ? "open" : "closed"]"
+		return
+	icon_state = "relay_on_[panel_open ? "open" : "closed"]"
+
+/obj/machinery/modular_shield/module/relay/RefreshParts()
+	. = ..()
+	range_boost = initial(range_boost)
+	for(var/datum/stock_part/micro_laser/new_laser in component_parts)
+		range_boost += new_laser.tier * 0.25
+
+	if(shield_generator)
+		shield_generator.calculate_boost()
+
+/obj/machinery/modular_shield/module/well
+
+	name = "Modular Shield Well"
+	desc = "A device used to hold more hardlight for the modular shield generator."
+	icon = 'icons/obj/machines/modular_shield_generator.dmi'
+	icon_state = "well_off_closed"
+
+	circuit = /obj/item/circuitboard/machine/modular_shield_well
+
+	///Amount of max strength this machine grants the connected generator
+	var/strength_boost = 0
+
+/obj/machinery/modular_shield/module/well/RefreshParts()
+	. = ..()
+	strength_boost = initial(strength_boost)
+	for(var/datum/stock_part/capacitor/new_capacitor in component_parts)
+		strength_boost += new_capacitor.tier * 10
+
+	if(shield_generator)
+		shield_generator.calculate_boost()
+
+/obj/machinery/modular_shield/module/well/update_icon_state()
+	. = ..()
+	if(isnull(shield_generator) || (machine_stat & NOPOWER))
+		icon_state = "well_off_[panel_open ? "open" : "closed"]"
+		return
+	icon_state = "well_on_[panel_open ? "open" : "closed"]"
+
+
+//The shield itself
+/obj/structure/emergency_shield/modular
+	name = "Modular energy shield"
+	desc = "An energy shield with varying configurations."
+	color = "#00ffff"
+	density = FALSE
+	alpha = 100
+	resistance_flags = INDESTRUCTIBLE //the shield itself is indestructible or atleast should be
+	//our parent
```

Comment:
Parent what?

Reply:
changed to ///the shield generator sustaining us